File: Arrays/Trapping_Rain_Water/TrappingRainWater.java
class TrappingRainWater{
    public int trap(int[] height){
        int left=0;
        int right=height.length-1;
        int leftMax=0,rightMax=0;
        int water=0;
        while(left<right){
            if(height[left]<height[right]){
                    ++left;
                    leftMax=Math.max(leftMax,height[left]);
                    water+=leftMax-height[left];
            }

            else{
                --right;
                rightMax=Math.max(rightMax,height[right]);
                water+=rightMax-height[right];
            }
        }   return water;
    }

    public static void main(String[] args) {
        TrappingRainWater trw=new TrappingRainWater();
        int[] height={0,1,0,2,1,0,1,3,2,1,2,1};
        System.out.println(trw.trap(height));
    }
}

File: Arrays/Trapping_Rain_Water/Trapping_Rain_Water.py
def trap(height) -> int:
    if not height:
        return 0

    left=0
    right =  len(height) - 1
    left_max, right_max = height[left], height[right]
    water_trapped = 0

    while left < right:
        if left_max < right_max:
            left += 1
            left_max = max(left_max, height[left])
            water_trapped += left_max - height[left]
        else:
            right -= 1
            right_max = max(right_max, height[right])
            water_trapped += right_max - height[right]

    return water_trapped
def main():
    height = [4,2,0,3,2,5]
    print(trap(height))
if __name__ == "__main__":
    main()
